File: src/main/java/nl/debo/cryptodata/Indicators.java
package nl.debo.cryptodata;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class Indicators {

    private Indicators() {
    }

    public static List<Double> rsi(
            List<Double> prices,
            int period
    ) {

        if (prices.size() <= period) {
            return Collections.emptyList();
        }

        var result = new ArrayList<Double>(
                Collections.nCopies(prices.size(), Double.NaN)
        );

        double gainSum = 0;
        double lossSum = 0;

        // Initial average gain/loss
        for (int i = 1; i <= period; i++) {

            double change = prices.get(i) - prices.get(i - 1);

            if (change > 0) {
                gainSum += change;
            } else {
                lossSum -= change;
            }
        }

        double averageGain = gainSum / period;
        double averageLoss = lossSum / period;

        result.set(
                period,
                calculateRsi(averageGain, averageLoss)
        );

        // Wilder smoothing
        for (int i = period + 1; i < prices.size(); i++) {

            double change =
                    prices.get(i) - prices.get(i - 1);

            double gain = Math.max(change, 0);
            double loss = Math.max(-change, 0);

            averageGain =
                    ((averageGain * (period - 1)) + gain)
                            / period;

            averageLoss =
                    ((averageLoss * (period - 1)) + loss)
                            / period;

            result.set(
                    i,
                    calculateRsi(
                            averageGain,
                            averageLoss
                    )
            );
        }

        return result;
    }

    private static double calculateRsi(
            double averageGain,
            double averageLoss
    ) {

        if (averageLoss == 0) {
            return 100.0;
        }

        if (averageGain == 0) {
            return 0.0;
        }

        double rs = averageGain / averageLoss;

        return 100.0 - (100.0 / (1.0 + rs));
    }

    public static List<Double> stochasticRsi(
            List<Double> rsi,
            int period
    ) {

        var result = new ArrayList<Double>(
                Collections.nCopies(rsi.size(), Double.NaN)
        );

        for (int i = period - 1; i < rsi.size(); i++) {

            double current = rsi.get(i);

            if (Double.isNaN(current)) {
                continue;
            }

            double lowest = Double.POSITIVE_INFINITY;
            double highest = Double.NEGATIVE_INFINITY;

            boolean valid = true;

            for (int j = i - period + 1; j <= i; j++) {

                double value = rsi.get(j);

                if (Double.isNaN(value)) {
                    valid = false;
                    break;
                }

                lowest = Math.min(lowest, value);
                highest = Math.max(highest, value);
            }

            if (!valid) {
                continue;
            }

            double range = highest - lowest;

            if (range == 0) {
                result.set(i, 0.0);
            } else {
                result.set(
                        i,
                        (current - lowest) / range
                );
            }
        }

        return result;
    }

    public static List<Double> ema(
            List<Double> values,
            int period
    ) {

        var result = new ArrayList<Double>(
                Collections.nCopies(values.size(), Double.NaN)
        );

        // Find the first index with `period` consecutive valid values and
        // seed the EMA with their simple average.
        int start = -1;
        int run = 0;

        for (int i = 0; i < values.size(); i++) {
            if (Double.isNaN(values.get(i))) {
                run = 0;
            } else {
                run++;
                if (run == period) {
                    start = i;
                    break;
                }
            }
        }

        if (start < 0) {
            return result;
        }

        double sum = 0;

        for (int j = start - period + 1; j <= start; j++) {
            sum += values.get(j);
        }

        double ema = sum / period;

        result.set(start, ema);

        double multiplier = 2.0 / (period + 1);

        for (int i = start + 1; i < values.size(); i++) {

            double value = values.get(i);

            if (Double.isNaN(value)) {
                continue;
            }

            ema = ((value - ema) * multiplier) + ema;

            result.set(i, ema);
        }

        return result;
    }

    public static List<Double> macdLine(
            List<Double> prices,
            int fastPeriod,
            int slowPeriod
    ) {

        var fast = ema(prices, fastPeriod);
        var slow = ema(prices, slowPeriod);

        var result = new ArrayList<Double>(
                Collections.nCopies(prices.size(), Double.NaN)
        );

        for (int i = 0; i < prices.size(); i++) {

            double f = fast.get(i);
            double s = slow.get(i);

            if (!Double.isNaN(f) && !Double.isNaN(s)) {
                result.set(i, f - s);
            }
        }

        return result;
    }

    public static List<Double> sma(
            List<Double> values,
            int period
    ) {

        var result = new ArrayList<Double>(
                Collections.nCopies(values.size(), Double.NaN)
        );

        double sum = 0;

        for (int i = 0; i < values.size(); i++) {

            double value = values.get(i);

            if (Double.isNaN(value)) {
                continue;
            }

            sum += value;

            if (i >= period) {

                double oldValue = values.get(i - period);

                if (!Double.isNaN(oldValue)) {
                    sum -= oldValue;
                }
            }

            if (i >= period - 1) {

                boolean valid = true;

                for (int j = i - period + 1; j <= i; j++) {
                    if (Double.isNaN(values.get(j))) {
                        valid = false;
                        break;
                    }
                }

                if (valid) {
                    result.set(i, sum / period);
                }
            }
        }

        return result;
    }
}

File: webapp/src/indicators.js
// Straight ports of nl.debo.cryptodata.tools.Indicators (Java). Every function
// returns an array of the same length as its input, NaN-padded during warmup.

const nanArray = (n) => new Array(n).fill(NaN);

/** RSI, Wilder smoothing (see rsi.md). First value at index `period`. */
export function rsi(prices, period = 14) {
  const result = nanArray(prices.length);
  if (prices.length <= period) return result;

  let gainSum = 0;
  let lossSum = 0;
  for (let i = 1; i <= period; i++) {
    const change = prices[i] - prices[i - 1];
    if (change > 0) gainSum += change;
    else lossSum -= change;
  }
  let avgGain = gainSum / period;
  let avgLoss = lossSum / period;
  result[period] = toRsi(avgGain, avgLoss);

  for (let i = period + 1; i < prices.length; i++) {
    const change = prices[i] - prices[i - 1];
    avgGain = (avgGain * (period - 1) + Math.max(change, 0)) / period;
    avgLoss = (avgLoss * (period - 1) + Math.max(-change, 0)) / period;
    result[i] = toRsi(avgGain, avgLoss);
  }
  return result;
}

function toRsi(avgGain, avgLoss) {
  if (avgLoss === 0) return 100;
  if (avgGain === 0) return 0;
  return 100 - 100 / (1 + avgGain / avgLoss);
}

/** Position of RSI inside its own min-max window, 0..1 (see stochastic-rsi.md). */
export function stochasticRsi(rsiSeries, period = 14) {
  const result = nanArray(rsiSeries.length);
  for (let i = period - 1; i < rsiSeries.length; i++) {
    const current = rsiSeries[i];
    if (Number.isNaN(current)) continue;
    let lowest = Infinity;
    let highest = -Infinity;
    let valid = true;
    for (let j = i - period + 1; j <= i; j++) {
      const v = rsiSeries[j];
      if (Number.isNaN(v)) { valid = false; break; }
      lowest = Math.min(lowest, v);
      highest = Math.max(highest, v);
    }
    if (!valid) continue;
    const range = highest - lowest;
    result[i] = range === 0 ? 0 : (current - lowest) / range;
  }
  return result;
}

/** NaN-aware simple moving average: only set when the whole window is valid. */
export function sma(values, period) {
  const result = nanArray(values.length);
  for (let i = period - 1; i < values.length; i++) {
    let sum = 0;
    let valid = true;
    for (let j = i - period + 1; j <= i; j++) {
      const v = values[j];
      if (Number.isNaN(v)) { valid = false; break; }
      sum += v;
    }
    if (valid) result[i] = sum / period;
  }
  return result;
}

/** EMA seeded on the first run of `period` consecutive valid values; NaN inputs are skipped. */
export function ema(values, period) {
  const result = nanArray(values.length);
  let start = -1;
  let run = 0;
  for (let i = 0; i < values.length; i++) {
    if (Number.isNaN(values[i])) run = 0;
    else if (++run === period) { start = i; break; }
  }
  if (start < 0) return result;

  let sum = 0;
  for (let j = start - period + 1; j <= start; j++) sum += values[j];
  let e = sum / period;
  result[start] = e;

  const k = 2 / (period + 1);
  for (let i = start + 1; i < values.length; i++) {
    const v = values[i];
    if (Number.isNaN(v)) continue;
    e = (v - e) * k + e;
    result[i] = e;
  }
  return result;
}

/** Average True Range, Wilder smoothing (port of the Java atr). First value at index `period`. */
export function atr(highs, lows, closes, period = 14) {
  const result = nanArray(closes.length);
  if (closes.length <= period) return result;
  const tr = (i) => Math.max(
    highs[i] - lows[i],
    Math.abs(highs[i] - closes[i - 1]),
    Math.abs(lows[i] - closes[i - 1]),
  );
  let sum = 0;
  for (let i = 1; i <= period; i++) sum += tr(i);
  let a = sum / period;
  result[period] = a;
  for (let i = period + 1; i < closes.length; i++) {
    a = (a * (period - 1) + tr(i)) / period;
    result[i] = a;
  }
  return result;
}

/** Bollinger Bands: SMA(period) ± mult population standard deviations. */
export function bollingerBands(prices, period = 20, mult = 2) {
  const middle = sma(prices, period);
  const upper = nanArray(prices.length);
  const lower = nanArray(prices.length);
  for (let i = 0; i < prices.length; i++) {
    if (Number.isNaN(middle[i])) continue;
    let sq = 0;
    for (let j = i - period + 1; j <= i; j++) sq += (prices[j] - middle[i]) ** 2;
    const sd = Math.sqrt(sq / period);
    upper[i] = middle[i] + mult * sd;
    lower[i] = middle[i] - mult * sd;
  }
  return { middle, upper, lower };
}

/** MACD line = EMA(fast) − EMA(slow); NaN unless both are valid. */
export function macdLine(prices, fast = 12, slow = 26) {
  const f = ema(prices, fast);
  const s = ema(prices, slow);
  return prices.map((_, i) =>
    Number.isNaN(f[i]) || Number.isNaN(s[i]) ? NaN : f[i] - s[i]
  );
}

/** Everything the chart needs, computed once per candle series. Same parameters as CryptoAnalysis. */
export function computeAll(candles) {
  const closes = candles.map((c) => c.close);
  const rsiS = rsi(closes, 14);
  const stoch = stochasticRsi(rsiS, 14);
  const k = sma(stoch, 3);
  const d = sma(k, 3);
  const macd = macdLine(closes, 12, 26);
  const signal = ema(macd, 9);
  const hist = macd.map((m, i) =>
    Number.isNaN(m) || Number.isNaN(signal[i]) ? NaN : m - signal[i]
  );
  const bb = bollingerBands(closes, 20, 2);
  return { closes, rsi: rsiS, stochRsi: stoch, k, d, macd, signal, hist, bbUpper: bb.upper, bbMiddle: bb.middle, bbLower: bb.lower };
}
